docs/developers/advanced_engines.rst

```diff
@@ -13,4 +12,54 @@ that's your interest (or if you're interested in methods like *ab initio* or
 quantum dynamics which might make use of a definition of state beyond
 standard MD), then this is the section for you.
 
+Standard snapshot features
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adding features with new stored data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adding features that are proxies to stored data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adding features that are properties
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Support for additional OPS modules
+----------------------------------
+
+The basic engine setup is easy to do, and enables the core of
+OpenPathSampling's functionality. However, some additional functionality can
+be added with very little effort, for engines where this might be relevant. 
+
+Shooting Point Velocity Modification
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The original descriptions of TPS used two-way shooting with a ":math:`\Delta
+p`" velocity modification. That is, the direction of the velocity on one
+particle was changed.
+
+OPS has the ability to do this using the
+:class:`.VelocityDirectionModifier`, but this procedure becomes extremely
+complicated in condensed phase (periodic) systems with holonomic
+constraints. Therefore, OPS only allows this for engines where the snapshots
+have the ``n_degrees_of_freedom`` feature. In addition, the engine can tell
+OPS not to bother removing linear momentum (e.g., if the engine represents a
+particle rolling on a surface, like our toy engine) by setting
+``engine.ignore_linear_momentum = True``.
+
+MDTraj support
+~~~~~~~~~~~~~~
+
+Several aspects of `MDTraj <http://mdtraj.org>`_ support require an
+:class:`mdtraj.Topology` object. The developer can add these by adding a
+property/attribute called ``mdtraj_topology`` to their engine. With this,
+you automatically get:
+
+* trajectory conversion to :class:`mdtraj.Trajectory`, which enables output
+  to various file formats, and integration with other projects such as
+  nglview for trajectory visualization.
+* ??? TODO: can we make MDTraj CV support automated as well? Check for topol
```

Comment:
I need to think about this. But it definitely might break immutability, since the actual CV depends on how you first call it. It would basically mean, that you define the CV independent of the topology. That might work. So the CV always grabs the topology of trajectory when it is used the first time.
We could really do that. But it might require a different CV class to not conflict with existing ones.

Reply:
> It would basically mean, that you define the CV independent of the topology.

Actually, I think this is okay. Compare with `FunctionCV`. If I say "give me the x-coordinate of atom 5", this CV will give different (semantic) results depending on what atom 5 is (i.e., change in topology). Same if I say "use MDTraj to give me the dihedral of atoms 11, 12, 13, 14" -- result might be nonsense for some other topology, but that's user error.

> But it might require a different CV class to not conflict with existing ones.

I was thinking that we could do it by keeping the existing `topology` argument, but making it optional. If `self.topology is None`, we set it when we first run. That maintains API compatibility.

Not sure the storage details here; we may need a special case when `topology` is `None`.

Reply:
> I was thinking that we could do it by keeping the existing topology argument, but making it optional. If self.topology is None, we set it when we first run. That maintains API compatibility.

Yes, that should work. So we change the documentation and agree that if you create a CV without topology this means that it works with the implicit defined topology. I think just using the first one and keep it might be problematic, but with a little simple logic that should work, too. We could cache it if the same engine is used.

Reply:
> We could cache it if the same engine is used.

Yep, I think that's the best approach. Something like this:

```python
if not self._cached_topology:
   self._cached_engine = snapshot.engine
   self._cached_topology = snapshot.mdtraj_topology

if snapshot.engine is self._cached_engine:
   topology = self._cached_topology
else:
   topology = snapshot.mdtraj_topology
```